### ISSUE TITLE
[C++][Integration] Refactor json_internal.cc to use fewer `DCHECK`s

### ISSUE THREAD

**Issue by bkietz** [apache/arrow#38186]
`json_internal.cc` uses `DCHECK`s to check many conditions even in scopes where a status could be returned instead. These should be replaced with error statuses, which will help pinpoint invalid JSON more explicitly

**Comment by bkietz** [comment]
No, I would say part of this PR would be refactoring UnboxValue to return a failing `Result` if the value were not the expected JSON type.

**Comment by felipecrv** [comment]
> No, I would say part of this PR would be refactoring UnboxValue to return a failing `Result` if the value were not the expected JSON type.

Isn't that going to add considerable overhead? The point of these `DCHECKs` is documenting the pre-conditions of these functions. In release builds we should test the type only on the caller, not on caller and on the `ARROW_CHECK`.

Perhaps we should document them and their callers with proper `/// \pre ...` Doxygen pre-conditions and fix any code violating the pre-conditions.

**Comment by bkietz** [comment]
I would guess that the overhead of parsing JSON would greatly outweigh the overhead of performing these checks. Also I believe this code isn't performance critical; these are only used during integration testing where datasets tend to be quite small.

**Comment by felipecrv** [comment]
> I would guess that the overhead of parsing JSON would greatly outweigh the overhead of performing these checks. Also I believe this code isn't performance critical; these are only used during integration testing where datasets tend to be quite small.

If it's just for tests, it's fair.

For production code, it's part of the contract of JSON parsing libraries that (since there's no schema) you only unbox when you already know the type.

**Comment by bkietz** [comment]
Right, which is why the JSON dataset reader skips RapidJSON's document/Object/Value API altogether.
